homeassistant/components/vacuum/mqtt.py

```diff
@@ -0,0 +1,455 @@
+            payload_stop, payload_clean_spot, payload_locate,
+            payload_start_pause, battery_level_topic, battery_level_template,
+            charging_topic, charging_template, state_topic, state_template,
+            fan_speed_topic, fan_speed_template, set_fan_speed_topic,
+            fan_speed_list, send_command_topic):
+        """Initialize the vacuum."""
+        self._name = name
+        self._supported_features = supported_features
+        self._qos = qos
+        self._retain = retain
+
+        self._command_topic = command_topic
+        self._payload_turn_on = payload_turn_on
+        self._payload_turn_off = payload_turn_off
+        self._payload_return_to_base = payload_return_to_base
+        self._payload_stop = payload_stop
+        self._payload_clean_spot = payload_clean_spot
+        self._payload_locate = payload_locate
+        self._payload_start_pause = payload_start_pause
+
+        self._battery_level_topic = battery_level_topic
+        self._battery_level_template = battery_level_template
+
+        self._charging_topic = charging_topic
+        self._charging_template = charging_template
+
+        self._state_topic = state_topic
+        self._state_template = state_template
+
+        self._fan_speed_topic = fan_speed_topic
+        self._fan_speed_template = fan_speed_template
+
+        self._set_fan_speed_topic = set_fan_speed_topic
+        self._fan_speed_list = fan_speed_list
+        self._send_command_topic = send_command_topic
+
+        self._is_cleaning = False
+        self._charging = False
+        self._status = 'Unknown'
+        self._battery_level = 0
+        self._fan_speed = 'unknown'
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Subscribe MQTT events.
+
+        This method is a coroutine.
+        """
+        @callback
+        def message_received(topic, payload, qos):
+            """Handle new MQTT message."""
+
+            if topic == self._battery_level_topic:
+                battery_level = self._battery_level_template\
+                    .async_render_with_possible_json_value(
+                        payload,
+                        error_value=None)
+                if battery_level is not None:
+                    self._battery_level = int(battery_level)
+
+            if topic == self._charging_topic:
+                charging = self._charging_template\
+                    .async_render_with_possible_json_value(
+                        payload,
+                        error_value=None)
+                if charging is not None:
+                    self._charging = charging
+
+            if topic == self._state_topic:
+                state = self._state_template\
+                    .async_render_with_possible_json_value(
+                        payload,
+                        error_value=None)
+                if state:
+                    if state == 'cleaning':
```

Comment:
I'm open to ideas on how to handle the `state` enum in a generic way as well.

Reply:
I wonder if it'd be better to just handle state as a set of bools potentially with their own topics/templates, then determine a state on the home assistant side.